state-chain/pallets/cf-threshold-signature/src/migrations/v4.rs

```diff
@@ -0,0 +1,99 @@
+use cf_traits::{Chainflip, EpochInfo};
+use frame_support::traits::OnRuntimeUpgrade;
+use sp_std::{marker::PhantomData, prelude::*};
+
+#[cfg(feature = "try-runtime")]
+mod try_runtime_includes {
+	pub use codec::{Decode, DecodeLength, Encode};
+	pub use frame_support::{
+		ensure, pallet_prelude::DispatchError, storage, traits::PalletInfoAccess,
+	};
+}
+#[cfg(feature = "try-runtime")]
+use try_runtime_includes::*;
+
+use crate::{CurrentKeyEpoch, KeyRotationStatus, PendingKeyRotation};
+
+mod old {
+	use cf_primitives::EpochIndex;
+	use frame_support::{storage_alias, Blake2_128Concat};
+
+	use crate::{AggKeyFor, Config, Pallet};
+
+	#[storage_alias]
+	pub type Vaults<T: Config<I>, I: 'static> =
+		StorageMap<Pallet<T, I>, Blake2_128Concat, EpochIndex, AggKeyFor<T, I>>;
+}
+
+/// The V4 migration is partly implemented in the runtime/lib.rs
+/// `ThresholdSignatureRefactorMigration` struct.
+pub struct Migration<T, I>(PhantomData<(T, I)>);
+
+impl<T: crate::Config<I>, I: 'static> OnRuntimeUpgrade for Migration<T, I> {
+	fn on_runtime_upgrade() -> frame_support::weights::Weight {
+		log::info!("Running V4 threshold migration");
+		// rename_pallet_storage::<crate::Pallet<T, I>>(b"Vaults", b"Keys");
+		for (k, v) in old::Vaults::<T, I>::drain() {
+			crate::Keys::<T, I>::insert(k, v);
+		}
+		CurrentKeyEpoch::<T, I>::put(<T as Chainflip>::EpochInfo::epoch_index());
```

Comment:
If we can get the epoch info here like this, is there even a reason we have `CurrentKeyEpoch` ?

Reply:
In some cases (during a rotation) they can get out of sync.